### PR TITLE
Fix spawn plugins not deleting temp checkpoint

### LINKED ISSUE #10933: `DDPSpawnPlugin` generates a file based on the "best model path"
## 🐛 Bug

In the DDPSpawn / TPUSpawn plugin we transfer the weights from rank 0 back to the main process. To do this, we save a checkpoint of the latest model weights and then load it in the main process. The file name is determined based on the checkpoint callback's best_model_path:

https://github.com/PyTorchLightning/pytorch-lightning/blob/a28b4cd0c0bba30c21cae571e650877f66cf5588/pytorch_lightning/plugins/training_type/ddp_spawn.py#L259-L261

This is not a bug that affects users directly as long as they ignore the file that's being saved. The name of the file does not reflect the state of the contents of that file, because the latest weights may not always be the best! 

Furthermore, the temp file never gets deleted. 

### To Reproduce

Run boring model with `Trainer(strategy="ddp_spawn", devices=2)`. The checkpoint directory will contain a file
`epoch=0-step=7.tmp_end.ckpt`

### Expected behavior

The filename is not based on the "best model path" and the file gets deleted after state has been loaded in main process.

### Additional context

Found during debugging in #10896. 
A PR for this fix is in the work.

cc @awaelchli @ananthsub @ninginthecloud @justusschock @kaushikb11

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #10933
A temporary checkpoint gets created when `Trainer(strategy="ddp_spawn")` but never gets deleted.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @awaelchli @ananthsub @ninginthecloud @justusschock @kaushikb11

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 1]
```diff
@@ -282,6 +282,7 @@ def __recover_results_in_main_process(self, spawn_output: "_SpawnOutput", traine
                 spawn_output.last_path, map_location=(lambda storage, loc: storage)
             )
             self.lightning_module.load_state_dict(ckpt)
+            self.checkpoint_io.remove_checkpoint(spawn_output.last_path)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:285]: reviewer: I'm adding the test for this in #10934 because it's easier that way.

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
